Use a TUN device and embedded PPP code, instead of using pppd

DimitriPapadopoulos opened the pull request:
This patch provided by @rain2fog in https://github.com/adrienverge/openfortivpn/issues/801#issuecomment-740306823 works for me. I get a working VPN tunnel without using `pppd`. I wouldn't mind help with reviewing.

From https://github.com/adrienverge/openfortivpn/issues/801#issuecomment-735513189:
> we simulated the PPP negotiation procedure. There are 3 stages in PPP protocal.
> * The 1th stage is LCP. client will send Magic-Number and MRU request to fortivpn server, then will ack the server request. server will send the Magic-Number and ack the client request.
> * The 2th stage is IPCP. both will will send the IPCP config-request to other. client need to send the IPCP config-request Option (3) with the IP which come from the XML. then both config-request need to be ack.
> * The 3th stage is 0x0021. we can read and write IP packet from tun NIC. then sending data by encapsulating or decapsulating the IP packet with header 0x0021(2 Bytes) + (totolsize(2 Bytes) + 0x5050 + pppsize(2 Bytes)).
we also need send the LCP echo-request to fortinet VPN to keep the TCP connection wouldn't close.

I get complier warnings on Ubuntu 22.04:
```
In function ‘conf_option_encode’,
    inlined from ‘conf_request’ at src/io.c:415:2:
src/io.c:337:16: warning: ‘request.tail’ may be used uninitialized [-Wmaybe-uninitialized]
  337 |         optlist->tail->type = type;
      |         ~~~~~~~^~~~~~
src/io.c: In function ‘conf_request’:
src/io.c:410:33: note: ‘request’ declared here
  410 |         struct conf_option_list request;
      |                                 ^~~~~~~
In function ‘conf_option_encode’,
    inlined from ‘ipcp_packet’ at src/io.c:984:4:
src/io.c:337:16: warning: ‘request.tail’ may be used uninitialized [-Wmaybe-uninitialized]
  337 |         optlist->tail->type = type;
      |         ~~~~~~~^~~~~~
src/io.c: In function ‘ipcp_packet’:
src/io.c:981:49: note: ‘request’ declared here
  981 |                         struct conf_option_list request;
      |                                                 ^~~~~~~
In function ‘conf_option_encode’,
    inlined from ‘ipcp_packet’ at src/io.c:1038:4:
src/io.c:337:16: warning: ‘request.tail’ may be used uninitialized [-Wmaybe-uninitialized]
  337 |         optlist->tail->type = type;
      |         ~~~~~~~^~~~~~
src/io.c: In function ‘ipcp_packet’:
src/io.c:1035:49: note: ‘request’ declared here
 1035 |                         struct conf_option_list request;
      |                                                 ^~~~~~~
```
Initially opened as #818.

Fixes #801.

DimitriPapadopoulos committed:
Apply patch provided by @rain2fog in #801

pppd → tun interface + embedded PPP code

DimitriPapadopoulos committed:
Clean up after applying patch

DimitriPapadopoulos committed:
Add patch provided by @rain2fog in #801

DimitriPapadopoulos committed:
When using a tun device, do not try to kill pppd

relaxlex commented:
Hi @DimitriPapadopoulos,

thank you for applying the TUN patch from @rain2fog. I have just ran a couple of short tests and this appears to work like a charm. Because I use openfortivpn daily I will use this tun branch build from now on to see how it performs. Because I can create a tun device as a normal user I will be running openfortivpn without sudo. 
Only difference I observed between the TUN and pppd version is that the TUN version is seeing a lot of 
"Route to gateway exists already." warnings when connecting to the same Forti VPN server.

Cheers!

relaxlex commented:
Hi @DimitriPapadopoulos,

thank you for the update, will test further with this rev. g2dec255 version. First glance TUN tunnel still working and _Route to gateway exists already_ warnings are also still there.

Cheers!

DimitriPapadopoulos commented:
Yes, I probably won't have time to look into the warnings any time soon. You are welcome to suggest a pull request to fix them.

DimitriPapadopoulos commented:
Fixed compiler warnings, just took a couple `assert` calls.

relaxlex commented:
Tested the rev. g2dec255 version for about 8 days and no issues so far, just as stable as the official 1.19.0  version. Today I switch to the new version in this branch, revision v1.20.1+git9.g47472c3, and will test this one as well.

Cheers,
-Lex

DimitriPapadopoulos commented:
Yes, connections are very stable for me too.

To merge this, we would need:
1. An option to switch from `pppd` to internal code. Id'l like to keep both, at least for a few releases, until we are certain that getting rid of `pppd` does not bring annoying side effects.
2. I don't have much time to look into the routing warnings and settings. If anyone has time to look into these, I can review and merge changes. Ideally, I'd like to switch from embedded C code to external scripts such as [vpnc-scripts](https://gitlab.com/openconnect/vpnc-scripts), but that can wait.

DimitriPapadopoulos committed:
Add experimental option --tun

Switch from pppd/ppp to a tun device and internal PPP code.

relaxlex commented:
Tested revision g9e9bdbe (1.20.1) from March 17th 2023 till June 21st 2023 without any issue. And now switched to revision gcde4ddd (1.20.4)

DimitriPapadopoulos commented:
I'm still working on the `--tun` option, I'll let you know when its ready for testing.

DimitriPapadopoulos commented:
It works for me with option `--tun=1`. We need to decide the best name for this option.

relaxlex commented:
Thank you @DimitriPapadopoulos for working on this. When I use the `--tun=1` argument a ppp interface is created. I think we also need to adjust the `merge_config` because after the merge function `cfg.tun` appears to be **0**, which would explain why a ppp interface is created and not a tun interface.  (Sorry not a programmer here, can read some code and put in some debug lines)

DblD committed:
Merge 'tun' config value if set in source configuration

This commit modifies the `merge_config` function to include the 'tun' value
from the source configuration if it is set. Previously, this value was not
being merged, leading to inconsistencies in the final configuration.

The change checks if 'tun' in the source configuration is different from its
default value before merging it into the destination configuration.

This ensures that the 'tun' setting is correctly propagated when
configurations are merged, allowing for more flexible and accurate VPN setups.

sino1641 commented:
Hi,
From @relaxlex it seems the experimental tun version should work. 
But I got some complie errors on the latest commit: 5493005

<details>
<summary>Log</summary>

```
/openfortivpn # git log -1 --format="%h"
5493005
/openfortivpn # make
  CC       src/openfortivpn-config.o
  CC       src/openfortivpn-hdlc.o
  CC       src/openfortivpn-http.o
  CC       src/openfortivpn-io.o
src/io.c: In function 'lcp_option_send':
src/io.c:384:43: error: 'PPP_LCP' undeclared (first use in this function); did you mean 'TUN_PPP_LCP'?
  384 |                         *ppp_type = htons(PPP_LCP);
      |                                           ^~~~~~~
      |                                           TUN_PPP_LCP
src/io.c:384:43: note: each undeclared identifier is reported only once for each function it appears in
src/io.c: In function 'lcp_packet':
src/io.c:457:38: error: 'PPP_CHAP' undeclared (first use in this function)
  457 |                                 case PPP_CHAP: {
      |                                      ^~~~~~~~
src/io.c: In function 'ipcp_add_route':
src/io.c:743:15: warning: implicit declaration of function 'ioctl' [-Wimplicit-function-declaration]
  743 |         ret = ioctl(sd, SIOCADDRT, &rt);
      |               ^~~~~
src/io.c:743:25: error: 'SIOCADDRT' undeclared (first use in this function); did you mean 'BIO_ADDR'?
  743 |         ret = ioctl(sd, SIOCADDRT, &rt);
      |                         ^~~~~~~~~
      |                         BIO_ADDR
src/io.c: In function 'ipcp_option_send':
src/io.c:770:43: error: 'PPP_IPCP' undeclared (first use in this function); did you mean 'TUN_PPP_IPCP'?
  770 |                         *ppp_type = htons(PPP_IPCP);
      |                                           ^~~~~~~~
      |                                           TUN_PPP_IPCP
src/io.c: In function 'pppd_write':
src/io.c:1245:30: error: 'PPP_LCP' undeclared (first use in this function); did you mean 'TUN_PPP_LCP'?
 1245 |                         case PPP_LCP:
      |                              ^~~~~~~
      |                              TUN_PPP_LCP
src/io.c:1248:30: error: 'PPP_IPCP' undeclared (first use in this function); did you mean 'TUN_PPP_IPCP'?
 1248 |                         case PPP_IPCP:
      |                              ^~~~~~~~
      |                              TUN_PPP_IPCP
src/io.c:1251:30: error: 'PPP_IP' undeclared (first use in this function)
 1251 |                         case PPP_IP:
      |                              ^~~~~~
src/io.c:1252:30: error: 'PPP_IPV6' undeclared (first use in this function)
 1252 |                         case PPP_IPV6:
      |                              ^~~~~~~~
make: *** [Makefile:575: src/openfortivpn-io.o] Error 1
```

DimitriPapadopoulos commented:
@sino1641 Something went wrong with the build process. Restart building from scratch, from a fresh checkout.

sino1641 commented:
@DimitriPapadopoulos Thanks for pointing it out. 
I tried to build it under an alpine container on a Linux host with no ppp kernel.
(But it builds a ppp-based binary from master successfully.)
Entire process:
<details>
<summary>Log</summary>

```
/ # apk add --no-cache autoconf automake build-base ca-certificates curl git openssl-dev
/ # git clone -b tun https://github.com/adrienverge/openfortivpn.git
/ # cd openfortivpn/
/openfortivpn # ./autogen.sh
+ type autoconf
+ type automake
+ type aclocal
+ aclocal
+ autoconf
+ automake --add-missing
configure.ac:10: installing './compile'
configure.ac:7: installing './install-sh'
configure.ac:7: installing './missing'
Makefile.am: installing './depcomp'
+ echo 'now you can run ./configure && make to build openfortivpn'
now you can run ./configure && make to build openfortivpn
/openfortivpn # ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /bin/sed
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for git... /usr/bin/git
checking whether make supports nested variables... (cached) yes
checking for libssl >= 1.0.2 libcrypto >= 1.0.2... yes
checking for pthread_create in -lpthread... yes
checking for forkpty in -lutil... yes
checking for libsystemd... no
libsystemd not present
checking for arpa/inet.h... yes
checking for fcntl.h... yes
checking for getopt.h... yes
checking for ifaddrs.h... yes
checking for netdb.h... yes
checking for net/if.h... yes
checking for netinet/in.h... yes
checking for netinet/tcp.h... yes
checking for pthread.h... yes
checking for strings.h... (cached) yes
checking for sys/ioctl.h... yes
checking for syslog.h... yes
checking for sys/socket.h... yes
checking for sys/stat.h... (cached) yes
checking for sys/types.h... (cached) yes
checking for sys/wait.h... yes
checking for termios.h... yes
checking for unistd.h... (cached) yes
checking for net/route.h... yes
checking for libutil.h... no
checking for linux/if_ppp.h... no
checking for linux/if_tun.h... no
checking for mach/mach.h... no
checking for pty.h... yes
checking for semaphore.h... yes
checking for util.h... no
checking for net/if_var.h... no
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for working volatile... yes
checking for off_t... yes
checking for pid_t... yes
checking for size_t... yes
checking for ssize_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint8_t... yes
checking for struct termios... yes
checking for access... yes
checking for close... yes
checking for connect... yes
checking for execv... yes
checking for _exit... yes
checking for fcntl... yes
checking for fileno... yes
checking for forkpty... yes
checking for freeaddrinfo... yes
checking for freeifaddrs... yes
checking for gai_strerror... yes
checking for getaddrinfo... yes
checking for geteuid... yes
checking for getifaddrs... yes
checking for getopt_long... yes
checking for htons... yes
checking for inet_addr... yes
checking for inet_ntoa... yes
checking for ioctl... yes
checking for isatty... yes
checking for memmem... yes
checking for ntohs... yes
checking for open... yes
checking for openlog... yes
checking for pclose... yes
checking for popen... yes
checking for pthread_cancel... yes
checking for pthread_cond_init... yes
checking for pthread_cond_signal... yes
checking for pthread_cond_wait... yes
checking for pthread_create... yes
checking for pthread_join... yes
checking for pthread_mutexattr_init... yes
checking for pthread_mutex_destroy... yes
checking for pthread_mutex_init... yes
checking for pthread_mutex_lock... yes
checking for pthread_mutex_unlock... yes
checking for pthread_self... yes
checking for pthread_sigmask... yes
checking for read... yes
checking for select... yes
checking for sem_destroy... yes
checking for sem_init... yes
checking for sem_post... yes
checking for sem_wait... yes
checking for setenv... yes
checking for setsockopt... yes
checking for sigaddset... yes
checking for sigemptyset... yes
checking for sleep... yes
checking for socket... yes
checking for strcasecmp... yes
checking for strdup... yes
checking for strncasecmp... yes
checking for strsignal... yes
checking for strtok_r... yes
checking for syslog... yes
checking for tcgetattr... yes
checking for tcsetattr... yes
checking for usleep... yes
checking for vsyslog... yes
checking for waitpid... yes
checking for write... yes
checking for pthread_mutexattr_setrobust... yes
checking whether rtentry is available and has rt_dst... yes
checking for /proc/net/route... yes
checking for ppp... /usr/sbin/ppp
checking for /usr/sbin/ppp... no
checking for pppd... /usr/sbin/pppd
checking for /usr/sbin/pppd... no
configure: HAVE_USR_SBIN_PPP... 0
configure: HAVE_USR_SBIN_PPPD... 1
configure: LEGACY_PPPD... 0
configure: HAVE_PROC_NET_ROUTE... 1
configure: PPP_PATH... /usr/sbin/pppd
checking for resolvconf... DISABLED
configure: RESOLVCONF_PATH... DISABLED
configure: HAVE_RESOLVCONF... 0
configure: USE_RESOLVCONF... 0
configure: systemdsystemunitdir... 
configure: HAVE_SO_BINDTODEVICE... 1
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: executing depfiles commands
config.status: executing timestamp commands
/openfortivpn # make
  CC       src/openfortivpn-config.o
  CC       src/openfortivpn-hdlc.o
  CC       src/openfortivpn-http.o
  CC       src/openfortivpn-io.o
src/io.c: In function 'lcp_option_send':
src/io.c:384:43: error: 'PPP_LCP' undeclared (first use in this function); did you mean 'TUN_PPP_LCP'?
  384 |                         *ppp_type = htons(PPP_LCP);
      |                                           ^~~~~~~
      |                                           TUN_PPP_LCP
src/io.c:384:43: note: each undeclared identifier is reported only once for each function it appears in
src/io.c: In function 'lcp_packet':
src/io.c:457:38: error: 'PPP_CHAP' undeclared (first use in this function)
  457 |                                 case PPP_CHAP: {
      |                                      ^~~~~~~~
src/io.c: In function 'ipcp_add_route':
src/io.c:743:15: warning: implicit declaration of function 'ioctl' [-Wimplicit-function-declaration]
  743 |         ret = ioctl(sd, SIOCADDRT, &rt);
      |               ^~~~~
src/io.c:743:25: error: 'SIOCADDRT' undeclared (first use in this function); did you mean 'BIO_ADDR'?
  743 |         ret = ioctl(sd, SIOCADDRT, &rt);
      |                         ^~~~~~~~~
      |                         BIO_ADDR
src/io.c: In function 'ipcp_option_send':
src/io.c:770:43: error: 'PPP_IPCP' undeclared (first use in this function); did you mean 'TUN_PPP_IPCP'?
  770 |                         *ppp_type = htons(PPP_IPCP);
      |                                           ^~~~~~~~
      |                                           TUN_PPP_IPCP
src/io.c: In function 'pppd_write':
src/io.c:1245:30: error: 'PPP_LCP' undeclared (first use in this function); did you mean 'TUN_PPP_LCP'?
 1245 |                         case PPP_LCP:
      |                              ^~~~~~~
      |                              TUN_PPP_LCP
src/io.c:1248:30: error: 'PPP_IPCP' undeclared (first use in this function); did you mean 'TUN_PPP_IPCP'?
 1248 |                         case PPP_IPCP:
      |                              ^~~~~~~~
      |                              TUN_PPP_IPCP
src/io.c:1251:30: error: 'PPP_IP' undeclared (first use in this function)
 1251 |                         case PPP_IP:
      |                              ^~~~~~
src/io.c:1252:30: error: 'PPP_IPV6' undeclared (first use in this function)
 1252 |                         case PPP_IPV6:
      |                              ^~~~~~~~
make: *** [Makefile:575: src/openfortivpn-io.o] Error 1
```
</details>

DimitriPapadopoulos commented:
- I have fixed `src/io.c` to include `<sys/ioctl.h>`, which should fix part of the compiler errors.
- I have fixed `configure.ac` to test for `<net/if_ppp.h>` instead of `<linux/if_ppp.h>`.
- I have modified `src/io.c` to define the missing `PPP_*` macros if `<net/if_ppp.h>`. Is this last change useful to you? Said otherwise, is  `<net/if_ppp.h>` missing from your machine?

sino1641 commented:
Sincerely thanks Dimitri!
It seems I missed package `linux-headers` which contains `linux/if_ppp.h` and `linux/if_tun.h`.
But then there is another error saying:
```
src/io.c:38:10: fatal error: net/if_ppp.h: No such file or directory
   38 | #include <net/if_ppp.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:575: src/openfortivpn-io.o] Error 1
```
Which means yes `HAVE_LINUX_IF_PPP_H` is working.

Searching Internet, I see that's because some conflicts on musl which Alpine is using so that `<net/if_ppp.h>` is missing.
(Refer:https://gitlab.alpinelinux.org/alpine/aports/-/issues/9643)

https://github.com/adrienverge/openfortivpn/blob/15ee42a91378d5dcf5024da9be8c52e1a94ef196/src/io.c#L44
And for your latest fix, I think this should be a positive status check to define the missing PPP_* macros as follows when <net/if_ppp.h> is missing. 

Now it works with all your commits, especially 15ee42a.

DimitriPapadopoulos commented:
The `!` in `!HAVE_NET_IF_PPP_H` was a temporary test to see what happens when `<net/if_ppp.h>` is not available.

I understand we need to switch from `<net/if_ppp.h>` to `<linux/if_ppp.h>` to support the variety of C system libraries available on Linux systems. i'll change that.

caeies commented:
Hi @DimitriPapadopoulos , I recently upgrade to Trixie, and of course the pppd option is no more working (pppd >=2.5.0) and things broken around in my use case.

So I decide to give this branch a try to "check" if it was working.

First issue is that the tun option didn't seems to honor the --no-route option which is bad because our VPN is for local only ressources, not a global VPN as many user may have. So this version is completly messing the routing table, and at the end, when openfortivpn ends it let the target ip addr route set instead of cleaning it up ... It even looks sometime that default route is completly missing ... (but that's random issue). [Note: in my case I added a ifup.d scripts in pppd configuration with a specific pppVPN interface so that I was properly configuring the local routes here].

Second issue is that in our case the VPN ask for releasing ressources every 4h ... and then we get stuck due to (probably I didn't investigate enough yet) the first issue: no more way to route packet to the target, so no way to logout and the endless loop is started until forever, which kills completly and at 100% the computer target internet connectivity (not good when you are working remotely :/). It events lead to a kill -9 only exit way as the process is completly stuck in kernel call: [<0>] do_signal_stop+0x16d/0x220
[<0>] get_signal+0x235/0x880
[<0>] arch_do_signal_or_restart+0x3f/0x260
[<0>] syscall_exit_to_user_mode+0x1bc/0x210
[<0>] do_syscall_64+0x8e/0x190
[<0>] entry_SYSCALL_64_after_hwframe+0x76/0x7e


I'll try to look at these issue, but I don't have a lot's of time. If you already know some place to look at I'll take them.

Thanks for your help.

caeies commented on src/main.c:
```diff
@@ -686,6 +700,41 @@ int main(int argc, char *argv[])
 	merge_config(&cfg, &cli_cfg);
 	set_syslog(cfg.use_syslog);
 
+	if (cfg.tun) {
```
Shouldn't be !cfg.tun ? or I miss something ?

caeies replied:
Ah I get the point ok I missed the EXIT_FAILURE, sorry for the noise ...

DimitriPapadopoulos replied:
> and of course the pppd option is no more working (pppd >=2.5.0)

I don't see why. t shoudl work just fine with pppd >= 2.5.0.

caeies replied:
Yeah, I miss the purpose of this part, sorry for the noise.

caeies replied:
Ah, Now I know why the pppd >= 2.5.0 is not working ... It's for the same reason than the tun is not working too ... In fact I found few days ago a description of the issue but I didn't link it to what I've got on my side. Unfortunately  I can't find anymore where I read the full issue description to credit it and pin point to it.

So the BIG issue (for both cases) is that the accept-remote is using the REAL INTERNET IP of the fortigate client ... which means that the openfortivpn add a dummy routing rule like:
ip route
....
X.Y.Z.U dev pppVPN proto kernel scope link src 10.0.0.10
....
Which means that we are routing packet to the fortigate internet address through the tunnel itself ... and bam nothing is working anymore ....
If we allow to setup the routes, then openfortivpn for a route to the X.Y.Z.U through the default ethernet interface which preceed this dummy rules and saves things ... but if we don't allow the routes to be overridden, then we are stuck in an broken connectivity case ...

But I don't see yet where openfortivpn is setting the gw route through the vpn interface itself ... continuing digging ...

caeies replied:
ok see https://github.com/adrienverge/openfortivpn/pull/1048/files#r2303711113 for the next debugging step ...

caeies committed:
Tun: allow to rename interface.

 * NOTE: the tun options are not yet integrated in the config file.
 * adding the --tun-ifname option to allow choosing interface name to
 ease script integrations.

caeies commented on src/io.c:
```diff
@@ -172,6 +175,897 @@ static os_semaphore_t sem_pppd_ready;
+		}
+
+		tun_ifup(tunnel->tun_iface, ip_address, peer_address);
+		ipv4_set_tunnel_routes(tunnel);
```
This is breaking things (most notably the --set-routes=0|1 configuration). And it kills the default route restoration when openfortivpn exits. Please let the proper callbacks on_ppp_up and on_ppp_down do their job as needed.

caeies reviewed (commented):
review still in progress on my side.

caeies commented on src/tunnel.c:
```diff
@@ -159,6 +162,122 @@ static int on_ppp_if_down(struct tunnel *tunnel)
+		log_info("setup <%s> ip addr to %s\n", ifname, inet_ntoa(sin->sin_addr));
+	}
+
+	if (peer_addr != 0) {
+		struct sockaddr_in *sin = (struct sockaddr_in *)&ifreq.ifr_addr;
+
+		sin->sin_family = AF_INET;
+		sin->sin_port = 0;
+		sin->sin_addr.s_addr = peer_addr;
+		ret = ioctl(fd, SIOCSIFDSTADDR, &ifreq);
+		TUN_ASSERT(ret == 0, "ioctl set dst ifaddr");
+	}
```
THIS is IT !!! if peer_addr == remote IP GW then you are dead ... If I get it properly, that might explain why depending on user it works or not. It looks like if our administrator configures the fortigate VPN to return the internet/external IP of the GW in the peer_addr (aka ipppd-accept-remote like), then everything is broken, but if the administrator use the INTERNAL IP then it should works like a charm. What I don't know is HOW fortigate Product set up this and why such silly configuration is used ...

Now last step is too properly configure external routes through usual mechanism and slap our forti administrator so that he don't ask the sslvpn to route itself through itself ...

In my case for now I comment this by adding a 0 && but best would be to warn the user of the missconfiguration here if peer_addr == external ip addr ...

Hope this is clear.

caeies replied:
Ok Got the info, in fact everything was already present, but, digging step by step between pppd and openfortivpn shows that the issue was already known: 
with the `set-routes=1` option ...

```
DEBUG:  ip route show to X.Y.Z.X/255.255.255.255 dev pppVPN
DEBUG:  Removing wrong route to vpn server...
```

So the issue is already known and silently workarounded in openfortivpn ... in the case the route are set. If routes are not set, then we hit the wall again with pppd (and as the code is ~duplicated in tun ppp emulation we hit the same wall).

Ok I will see if I can do something about that properly.

@DimitriPapadopoulos , at least we have the full story of this annoying piece ...

caeies replied:
ok pppd cases fixed too. I'll check if I can create a specific MR for this.

caeies committed:
Adding a sample script for tun additionnal ressources configuration.

caeies committed:
Tun: allow the tun option in config file.

caeies commented:
@DimitriPapadopoulos : not sure this is how things should be done, but I opened a pull request on this branch to add some useful "features" ... see #1299

DimitriPapadopoulos commented:
Otherwise I should be able to restore the previous state of the branch from another computer in a few weeks.

caeies commented:
Here's the format-patch for 42f012319476890b2fe522ab0fb78f75d9fdfb56 @DimitriPapadopoulos 

[0001-Merge-tun-config-value-if-set-in-source-configuratio.patch](https://github.com/user-attachments/files/22022342/0001-Merge-tun-config-value-if-set-in-source-configuratio.patch)

caeies committed:
Tun: properly release resources at end of connection.

DimitriPapadopoulos committed:
Tun: fix typos found by codespell

DimitriPapadopoulos committed:
Fix Coverity defect

CID 488054 RESOURCE_LEAK
Overwriting handle `tunnel.pppd_pty` in `tunnel.pppd_pty = -1`
leaks the handle.

caeies commented:
Something is broken on the last code version. it drops the default route even if set-routes = 0 and failed to restore it on tunnel restoration ... maybe something broken with unitialized memory ...

DimitriPapadopoulos commented on src/tunnel.c:
```diff
@@ -1575,10 +1575,9 @@ int run_tunnel(struct vpn_config *config)
 		ret = pppd_terminate(&tunnel);
 		log_info("Terminated %s.\n", PPP_DAEMON);
 	} else {
-		if (tunnel.pppd_pty > 0)
```
No worries. I think there's still an issue in the initial code. I will let you know.

DimitriPapadopoulos commented:
There's one last Coverity Scan issue in the TUN code from [**rain2fog**](https://github.com/rain2fog):
https://github.com/adrienverge/openfortivpn/blob/93b89ca995c585bd5f59b0e0f9165e7db723a71c/src/http.c#L873-L878

CID 364257 Dereference after null check (FORWARD_NULL)
Passing null pointer gateway to `inet_addr`, which dereferences it.

@caeies It would be great if you could look into it. I am fixing other issues.
